flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1963,7 +2020,14 @@ public void startCheckpointScheduler() {
             stopCheckpointScheduler();
 
             periodicScheduling = true;
-            currentPeriodicTrigger = scheduleTriggerWithDelay(getRandomInitDelay());
+            if (baseIntervalDuringBacklog != baseInterval) {
```

Comment:
The code that enforces the checkpoint-interval-during-backlog is far more complicated than it should be. For example, it would not be necessary to have two parallel periodic tasks scheduled at the same time.

It should be much simpler and more readable to do the following:

```
- Use a variable to maintain the absolute time of last checkpoint.
- Let method `getNextCheckpointDelay()` return the time to wait from now until we should trigger the next checkpoint. T
- Let method `scheduleAndMaybeTriggerCheckpoint()` do the following:
  - Invoke `triggerCheckpoint()` if `getNextCheckpointDelay()` <= 0
  - Schedule `scheduleAndMaybeTriggerCheckpoint()` with delay=`getNextCheckpointDelay()`.
- `startCheckpointScheduler()` should schedule `scheduleAndMaybeTriggerCheckpointI()` with delay=`getNextCheckpointDelay()`
- Whenever isBacklog status changes, we should cancel the pending scheduled task and re-schedule it based on `getNextCheckpointDelay()`.
```